collector/shards.go

```diff
@@ -0,0 +1,177 @@
+// Copyright 2021 The Prometheus Authors
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+package collector
+
+import (
+	"encoding/json"
+	"fmt"
+	"net/http"
+	"net/url"
+	"path"
+
+	"github.com/go-kit/log"
+	"github.com/go-kit/log/level"
+	"github.com/prometheus/client_golang/prometheus"
+)
+
+var (
+	defaultNodeShardLabels = []string{"node"}
+
+	defaultNodeShardLabelValues = func(node string) []string {
+		return []string{
+			node,
+		}
+	}
+)
+
+// Shards information struct
+type Shards struct {
+	logger log.Logger
+	client *http.Client
+	url    *url.URL
+
+	nodeShardMetrics  []*nodeShardMetric
+	jsonParseFailures prometheus.Counter
+}
+
+// NodeShard Information per node struct
+type NodeShard struct {
+	node   string
+	shards int64
+}
+
+type nodeShardMetric struct {
+	Type   prometheus.ValueType
+	Desc   *prometheus.Desc
+	Value  func(shards float64) float64
+	Labels func(node string) []string
+}
+
+// NewShards defines Shards Prometheus metrics
+func NewShards(logger log.Logger, client *http.Client, url *url.URL) *Shards {
+	return &Shards{
+		logger: logger,
+		client: client,
+		url:    url,
+
+		nodeShardMetrics: []*nodeShardMetric{
+			{
+				Type: prometheus.GaugeValue,
+				Desc: prometheus.NewDesc(
+					prometheus.BuildFQName(namespace, "node_shards", "total"),
+					"Total shards per node",
+					defaultNodeShardLabels, nil,
+				),
+				Value: func(shards float64) float64 {
+					return shards
+				},
+				Labels: defaultNodeShardLabelValues,
+			}},
+
+		jsonParseFailures: prometheus.NewCounter(prometheus.CounterOpts{
+			Name: prometheus.BuildFQName(namespace, "node_shards", "json_parse_failures"),
+			Help: "Number of errors while parsing JSON.",
+		}),
+	}
+}
+
+// Describe Shards
+func (s *Shards) Describe(ch chan<- *prometheus.Desc) {
+	ch <- s.jsonParseFailures.Desc()
+
+	for _, metric := range s.nodeShardMetrics {
+		ch <- metric.Desc
+	}
+}
+
+func (s *Shards) getAndParseURL(u *url.URL, data interface{}) error {
```

Comment:
Instead of taking the data as a parameter, it should just return the response type. It seems odd to me to take the return value as an input parameter.